File: app/src/main/java/com/example/win/easy/viewmodel/SongViewModel.java
package com.example.win.easy.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.example.win.easy.value_object.VOUtil;
import com.example.win.easy.repository.db.data_object.SongDO;
import com.example.win.easy.repository.deprecated.repo.__SongRepository;
import com.example.win.easy.value_object.SongVO;

import java.util.ArrayList;
import java.util.List;

public class SongViewModel extends ViewModel {

    private __SongRepository songRepository;
    private VOUtil voUtil;

    public SongViewModel(__SongRepository songRepository,VOUtil voUtil){
        this.songRepository=songRepository;
        this.voUtil=voUtil;
    }

    private LiveData<List<SongVO>> allSongsLiveData;

    /**
     * 获取VO形式的所有歌曲
     * @return 所有歌曲的LiveData
     */
    public LiveData<List<SongVO>> getAllSongs(){
        //懒惰加载
        if (allSongsLiveData==null){
            //从repo获取DO形式的歌曲，并提供一个映射函数
            allSongsLiveData=Transformations.map(songRepository.getAll(),
                    //对每个DO依次转化
                    input -> {
                List<SongVO> songVOs=new ArrayList<>();
                for(SongDO songDO:input)
                    songVOs.add(voUtil.toVO(songDO));
                return  songVOs;
            });
        }
        return allSongsLiveData;
    }

}

File: app/src/main/java/com/example/win/easy/viewmodel/SongViewModel.kt
package com.example.win.easy.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.win.easy.repository.SongRepository
import com.example.win.easy.db.SongDO
import com.example.win.easy.db.SongListDO
import kotlinx.coroutines.launch
import kotlin.collections.ArrayList

class SongViewModel(
        private val songRepository: SongRepository) : ViewModel() {
    val allLocalSongs=songRepository.allLocalData
    val allSongs=songRepository.allData

    fun insertOnUniquePath(songDO: SongDO) =viewModelScope.launch {
        songRepository.insertOnUniquePath(songDO)
    }

    fun launchLoadSongsBySeq(sequence: List<Char>, block: (songs: List<SongDO>)->Unit){
        viewModelScope.launch {
            block(songRepository.loadSongsBySeq(sequence))
        }
    }

    fun launchLoadSongsExclude(songListDO: SongListDO, block: (songs: List<SongDO>) -> Unit){
        viewModelScope.launch {
            block(songRepository.loadSongsExclude(songListDO))
        }
    }

    fun launchLoadSongsBySongLists(songLists: List<SongListDO>, block: (songs: List<List<SongDO>>) -> Unit){
        viewModelScope.launch {
            ArrayList<List<SongDO>>().run {
                for (songList in songLists)
                    add(songRepository.loadSongsBySongList(songList))
                block(this)
            }
        }
    }
}